Refactoring RSTensor Share Distribution and Optimised unit tests

kamathhrishi opened the pull request:
## Description
This PR parallelizes RSTensor share distribution and optimizes unit tests by running for just enough parties to test functionality. 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)

gmuraru commented on tests/sympc/tensor/replicatedshare_tensor_test.py:
```diff
@@ -306,7 +306,7 @@ def test_ops_public_mul_integer(get_clients, parties, security):
     assert (result.reconstruct() == (secret * value)).all()
 
 
-@pytest.mark.parametrize("parties", [3, 5, 7])
```
Q: Those changes are for speed-up?

kamathhrishi replied:
Yeah for speed up. I have addressed this as a general comment. Let me know what you think.

LaRiffle commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -581,6 +581,43 @@ def reconstruct(
+        shares: List[Union[ShareTensor, torch.Tensor]],
+        party_rank: int,
+        session: Session,
+    ):
```
Add the return annotation type

LaRiffle commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -581,6 +581,43 @@ def reconstruct(
+            tensor (ReplicatedSharedTensor): Tensor with shares
+        """
+        party = session.parties[party_rank]
+        nshares = session.nr_parties - 1
```
I'd rather not introduce this variable which doesn't seem very useful

kamathhrishi replied:
I feel it makes the code more easily read rather than having just session.nr_parties-1 somewhere, someone reading the code would easily understand its the number of sharea.

LaRiffle commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -581,6 +581,43 @@ def reconstruct(
+                party_shares.append(share)
+
+            elif isinstance(share, ShareTensor):
+                party_shares.append(share.tensor)
```
Add a final `else` statement to raise an Typeerror

LaRiffle commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -605,32 +642,13 @@ def distribute_shares(
-            ptr_list.append(tensor)
-
-        return ptr_list
+        return parallel_execution(ReplicatedSharedTensor.distribute_shares_to_party)(
```
have you checked that parallel execution is useful here? It creates sub processes that can be slow to launch, so sometimes non parellel (ie classic) execution is faster. I would suggest running X experiments and comparing before implementing this feature

kamathhrishi replied:
Oh alright. Will have a look.

kamathhrishi replied:
Yeah, classic was slightly better, I will switch back to classic.

gmuraru replied:
Yeah we can switch to the older one - I do consider that we might revisit this and do it like you have here.
For the moment we always create a Thread/Process pool for parallelization - I think that if we use 3-4 parties (with VirtualMachines -- the testing setup) it would not give better results because we spend more time creating the thread pool than doing the effective work.

What might change, if we switch to real multi-party over the network, it might be that this approach might give better results depending on the number of parties.

kamathhrishi replied:
I will be keeping the refactoring part. That is having a separate function to distribute shares to a party and an overall function to distribute it to all parties. In the future would just add a parallel_execution statement around it.

LaRiffle commented on tests/sympc/tensor/mpc_tensor_test.py:
```diff
@@ -125,7 +125,7 @@ def test_local_secret_not_tensor(get_clients) -> None:
 
 
-@pytest.mark.parametrize("nr_clients", [2, 3, 4, 5])
+@pytest.mark.parametrize("nr_clients", [2, 5])
```
Agreed with @gmuraru, why do we change this?

kamathhrishi replied:
I have addressed this as a general comment. Let me know what you think.

gmuraru replied:
Let's also add the "3" value for those -- I think that 2, 3, 5 should suffice - 2 for FSS, 3 for Falcon, and 5 for a more general case.

LaRiffle reviewed (changes requested):
Some great syntax optimisations! I've left some comments

kamathhrishi commented:
@gmuraru @LaRiffle Those are changes for speedup. 
I think it wasn't right for me to assume what works for 2 or 5 parties will work the same for 3,4 or 7 parties for FSS and SPDZ.

I noticed that was the case for ReplicatedSharedTensor (RSTensor) operations implemented and removed tests I had written for 7&11 parties. In the case of RSTensor what works for 2 or 5 parties works well for say 3,4,...,n parties so having tests for 7 & 11 parties were piling up CI time and compute. Unit tests should test for correct functionality. 

After removing just tests for 7&11 parties I noticed time taken to run RSTensor tests reduced from ~800s to ~230s. 
MPCTensor from ~5500s to ~2300. In MPCTensors tests, I removed some tests that I thought were redundant like instead of having tests for 2,3,4,5, and just 2 & 5. 

For RStensor having just 2 & 5 parties (some case 3&5 or just 3) tests would work. I am not sure if functionality varies for different parties for SPDZ and FSS.

What do you think?

gmuraru commented on src/sympc/protocol/falcon/falcon.py:
```diff
@@ -98,18 +98,11 @@ def mul_master(x: MPCTensor, y: MPCTensor, session: Session) -> MPCTensor:
-                        y.share_ptrs[index],
-                    ]
-                )
+            args = [[x.share_ptrs[index], y.share_ptrs[index]] for index in range(3)]
```
MAYBE: `list(zip(x.share_ptrs, y.share_ptrs)`

gmuraru commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -78,8 +78,9 @@ def __init__(
-            for i in range(len(shares)):
-                self.shares.append(self._encode(shares[i]).to(tensor_type))
+            self.shares = [
+                self._encode(shares[i]).to(tensor_type) for i in range(len(shares))
```
MAYBE: `self._encode(share).to(tensor_type) for share in shares`

gmuraru commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -109,9 +110,8 @@ def _decode(self) -> List[torch.Tensor]:
-        for i in range(len(self.shares)):
-            tensor = self.fp_encoder.decode(self.shares[i].type(torch.LongTensor))
-            shares.append(tensor)
+        for share in self.shares:
```
Q: Maybe list comprehension also for this?

gmuraru commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -695,9 +716,8 @@ def method_new_rs_tensor(
         ) -> Any:
             shares = []
-            for i in range(len(_self.shares)):
-                tensor = getattr(_self.shares[i], method_name)(*args, **kwargs)
```
Q: Could we leave them as they were before since it might help  when we try to debug

kamathhrishi replied:
Makes sense.

gmuraru reviewed (approved):
Only some comments but in rest, it looks good!

rasswanth-s commented on src/sympc/tensor/replicatedshare_tensor.py:
```diff
@@ -581,18 +580,61 @@ def reconstruct(
         """Distribute a list of shares.
 
         Args:
             shares (List[ShareTensor): list of shares to distribute.
```
I think the it should also include torch.Tensor in args.

rasswanth-s reviewed (approved):
LGTM!
